File: prototype/src/protocol.ts
/**
 * Pgvpd — Postgres Wire Protocol Primitives
 *
 * Implements the minimum subset of the Postgres v3 wire protocol needed
 * for the proxy to parse StartupMessages, detect auth completion, inject
 * SET commands, and frame messages for transparent forwarding.
 *
 * Reference: https://www.postgresql.org/docs/current/protocol-message-formats.html
 */

// ─── Constants ──────────────────────────────────────────────────────────────

/** Postgres protocol version 3.0 */
export const PROTOCOL_VERSION_30 = 196608; // 0x00030000

/** SSLRequest magic number */
export const SSL_REQUEST_CODE = 80877103; // 0x04D2162F

/** CancelRequest magic number */
export const CANCEL_REQUEST_CODE = 80877102;

/** Single-byte response denying SSL */
export const SSL_DENY = Buffer.from([0x4e]); // 'N'

/** Backend message type bytes */
export const BackendMsg = {
  Authentication: 0x52, // 'R'
  ParameterStatus: 0x53, // 'S'
  BackendKeyData: 0x4b, // 'K'
  ReadyForQuery: 0x5a, // 'Z'
  CommandComplete: 0x43, // 'C'
  RowDescription: 0x54, // 'T'
  DataRow: 0x44, // 'D'
  ErrorResponse: 0x45, // 'E'
  NoticeResponse: 0x4e, // 'N'
  EmptyQueryResponse: 0x49, // 'I'
} as const;

/** Authentication subtypes (first Int32 after 'R' length) */
export const AuthType = {
  Ok: 0,
  CleartextPassword: 3,
  MD5Password: 5,
  SASL: 10,
  SASLContinue: 11,
  SASLFinal: 12,
} as const;

// ─── StartupMessage ─────────────────────────────────────────────────────────

export interface StartupParams {
  /** Key-value pairs from the startup message (user, database, etc.) */
  params: Map<string, string>;
}

/**
 * Parse a StartupMessage buffer into its parameters.
 *
 * StartupMessage format (no type byte):
 *   Int32 length (including itself)
 *   Int32 protocol version (196608 for 3.0)
 *   { String key \0 String value \0 }*
 *   \0  (terminal null)
 */
export function parseStartupMessage(buf: Buffer): StartupParams {
  const params = new Map<string, string>();
  // Skip length (4 bytes) and version (4 bytes)
  let offset = 8;
  const length = buf.readInt32BE(0);

  while (offset < length - 1) {
    // Read key
    const keyEnd = buf.indexOf(0, offset);
    if (keyEnd === -1 || keyEnd >= length) break;
    const key = buf.toString("utf8", offset, keyEnd);
    offset = keyEnd + 1;

    // Read value
    const valEnd = buf.indexOf(0, offset);
    if (valEnd === -1 || valEnd >= length) break;
    const value = buf.toString("utf8", offset, valEnd);
    offset = valEnd + 1;

    if (key) params.set(key, value);
  }

  return { params };
}

/**
 * Build a StartupMessage buffer from parameters.
 *
 * Reconstructs a valid StartupMessage with protocol version 3.0.
 */
export function buildStartupMessage(params: Map<string, string>): Buffer {
  // Calculate payload size: version + key-value pairs + terminal null
  let payloadLen = 4; // protocol version
  for (const [key, value] of params) {
    payloadLen += Buffer.byteLength(key, "utf8") + 1; // key + null
    payloadLen += Buffer.byteLength(value, "utf8") + 1; // value + null
  }
  payloadLen += 1; // terminal null

  const totalLen = 4 + payloadLen; // 4 for length field itself
  const buf = Buffer.alloc(totalLen);
  let offset = 0;

  buf.writeInt32BE(totalLen, offset);
  offset += 4;
  buf.writeInt32BE(PROTOCOL_VERSION_30, offset);
  offset += 4;

  for (const [key, value] of params) {
    offset += buf.write(key, offset, "utf8");
    buf[offset++] = 0;
    offset += buf.write(value, offset, "utf8");
    buf[offset++] = 0;
  }
  buf[offset] = 0; // terminal null

  return buf;
}

// ─── Query Message ──────────────────────────────────────────────────────────

/**
 * Build a SimpleQuery ('Q') message.
 *
 * Format: 'Q' Int32(length) String(query \0)
 */
export function buildQueryMessage(sql: string): Buffer {
  const sqlBytes = Buffer.from(sql, "utf8");
  const msgLen = 4 + sqlBytes.length + 1; // length field + sql + null terminator
  const buf = Buffer.alloc(1 + msgLen);

  buf[0] = 0x51; // 'Q'
  buf.writeInt32BE(msgLen, 1);
  sqlBytes.copy(buf, 5);
  buf[5 + sqlBytes.length] = 0; // null terminator

  return buf;
}

// ─── ErrorResponse ──────────────────────────────────────────────────────────

/**
 * Build an ErrorResponse ('E') message to send to the client.
 *
 * Format: 'E' Int32(length) { Byte1(fieldType) String(value \0) }* \0
 */
export function buildErrorResponse(
  severity: string,
  sqlstate: string,
  message: string,
  detail?: string,
): Buffer {
  const fields: Array<{ type: string; value: string }> = [
    { type: "S", value: severity },
    { type: "V", value: severity }, // non-localized severity
    { type: "C", value: sqlstate },
    { type: "M", value: message },
  ];
  if (detail) {
    fields.push({ type: "D", value: detail });
  }

  // Calculate total field bytes
  let fieldsLen = 0;
  for (const f of fields) {
    fieldsLen += 1 + Buffer.byteLength(f.value, "utf8") + 1; // type byte + value + null
  }
  fieldsLen += 1; // terminal null

  const msgLen = 4 + fieldsLen; // length includes itself
  const buf = Buffer.alloc(1 + msgLen);

  buf[0] = 0x45; // 'E'
  buf.writeInt32BE(msgLen, 1);

  let offset = 5;
  for (const f of fields) {
    buf[offset++] = f.type.charCodeAt(0);
    offset += buf.write(f.value, offset, "utf8");
    buf[offset++] = 0;
  }
  buf[offset] = 0; // terminal null

  return buf;
}

// ─── Message Framer ─────────────────────────────────────────────────────────

/**
 * Parsed backend message (server → client).
 */
export interface BackendMessage {
  /** Message type byte */
  type: number;
  /** Total message length (type byte + length field + payload) */
  totalLength: number;
  /** Raw payload after the length field */
  payload: Buffer;
  /** The complete raw message bytes (for forwarding) */
  raw: Buffer;
}

/**
 * Buffers TCP stream data and extracts complete Postgres protocol messages.
 *
 * TCP is a stream protocol — we may receive partial messages or multiple
 * messages in a single chunk. This class handles reassembly.
 */
export class MessageFramer {
  private buffer: Buffer = Buffer.alloc(0);

  /** Append new data from the TCP stream. */
  push(data: Buffer): void {
    this.buffer = Buffer.concat([this.buffer, data]);
  }

  /** Returns true if there's any buffered data. */
  hasData(): boolean {
    return this.buffer.length > 0;
  }

  /** Return the raw buffer (for forwarding remaining data). */
  remaining(): Buffer {
    return this.buffer;
  }

  /** Drain and return all buffered data, resetting the buffer. */
  drain(): Buffer {
    const data = this.buffer;
    this.buffer = Buffer.alloc(0);
    return data;
  }

  /**
   * Try to read the initial startup-phase message from a client.
   *
   * Startup messages have no type byte — they start with Int32 length.
   * This handles StartupMessage, SSLRequest, and CancelRequest.
   *
   * Returns null if not enough data yet.
   */
  nextStartupMessage(): Buffer | null {
    if (this.buffer.length < 4) return null;

    const length = this.buffer.readInt32BE(0);
    if (length < 4 || length > 10240) return null; // sanity check

    if (this.buffer.length < length) return null;

    const msg = Buffer.from(this.buffer.subarray(0, length));
    this.buffer = this.buffer.subarray(length);
    return msg;
  }

  /**
   * Try to read a complete backend message (server → client).
   *
   * Backend messages: Byte1(type) Int32(length) payload
   * The length field includes itself (4 bytes) but not the type byte.
   *
   * Returns null if not enough data yet.
   */
  nextBackendMessage(): BackendMessage | null {
    if (this.buffer.length < 5) return null;

    const type = this.buffer[0];
    const length = this.buffer.readInt32BE(1); // includes the 4 length bytes
    const totalLength = 1 + length; // type byte + length value

    if (this.buffer.length < totalLength) return null;

    const raw = Buffer.from(this.buffer.subarray(0, totalLength));
    const payload = Buffer.from(this.buffer.subarray(5, totalLength));
    this.buffer = this.buffer.subarray(totalLength);

    return { type, totalLength, payload, raw };
  }
}

// ─── Message Inspection Helpers ─────────────────────────────────────────────

/** Check if a startup-phase buffer is an SSLRequest. */
export function isSSLRequest(buf: Buffer): boolean {
  if (buf.length < 8) return false;
  return buf.readInt32BE(4) === SSL_REQUEST_CODE;
}

/** Check if a startup-phase buffer is a CancelRequest. */
export function isCancelRequest(buf: Buffer): boolean {
  if (buf.length < 8) return false;
  return buf.readInt32BE(4) === CANCEL_REQUEST_CODE;
}

/** Check if a backend message is AuthenticationOk. */
export function isAuthOk(msg: BackendMessage): boolean {
  return (
    msg.type === BackendMsg.Authentication &&
    msg.payload.length >= 4 &&
    msg.payload.readInt32BE(0) === AuthType.Ok
  );
}

/** Check if a backend message is ReadyForQuery. */
export function isReadyForQuery(msg: BackendMessage): boolean {
  return msg.type === BackendMsg.ReadyForQuery;
}

/** Check if a backend message is an ErrorResponse. */
export function isErrorResponse(msg: BackendMessage): boolean {
  return msg.type === BackendMsg.ErrorResponse;
}

/**
 * Extract human-readable error text from an ErrorResponse message.
 */
export function extractErrorMessage(msg: BackendMessage): string {
  const parts: string[] = [];
  let offset = 0;

  while (offset < msg.payload.length) {
    const fieldType = msg.payload[offset];
    if (fieldType === 0) break; // terminal null
    offset++;

    const strEnd = msg.payload.indexOf(0, offset);
    if (strEnd === -1) break;

    const value = msg.payload.toString("utf8", offset, strEnd);
    offset = strEnd + 1;

    if (fieldType === 0x4d) {
      // 'M' = Message
      parts.unshift(value);
    } else if (fieldType === 0x44) {
      // 'D' = Detail
      parts.push(value);
    }
  }

  return parts.join(": ") || "unknown error";
}

File: src/protocol.rs
//! Postgres Wire Protocol Primitives
//!
//! Minimum subset of the Postgres v3 wire protocol needed for the proxy:
//! parse StartupMessages, detect auth completion, inject SET commands,
//! frame messages for transparent forwarding.
//!
//! Reference: https://www.postgresql.org/docs/current/protocol-message-formats.html

use bytes::{Buf, BufMut, BytesMut};
use std::collections::HashMap;
use std::io;

// ─── Constants ──────────────────────────────────────────────────────────────

/// Postgres protocol version 3.0
const PROTOCOL_VERSION_30: i32 = 196608; // 0x00030000

/// SSLRequest magic number
const SSL_REQUEST_CODE: i32 = 80877103;

/// CancelRequest magic number
const CANCEL_REQUEST_CODE: i32 = 80877102;

/// Single byte denying SSL
pub const SSL_DENY: &[u8] = &[b'N'];

/// Backend message types we care about
pub mod backend {
    pub const AUTHENTICATION: u8 = b'R';
    pub const PARAMETER_STATUS: u8 = b'S';
    pub const BACKEND_KEY_DATA: u8 = b'K';
    pub const READY_FOR_QUERY: u8 = b'Z';
    pub const COMMAND_COMPLETE: u8 = b'C';
    pub const ERROR_RESPONSE: u8 = b'E';
    pub const ROW_DESCRIPTION: u8 = b'T';
    pub const DATA_ROW: u8 = b'D';
    pub const EMPTY_QUERY_RESPONSE: u8 = b'I';
}

/// Authentication subtypes
pub mod auth {
    pub const OK: i32 = 0;
    pub const CLEARTEXT_PASSWORD: i32 = 3;
    pub const MD5_PASSWORD: i32 = 5;
    pub const SASL: i32 = 10;
    pub const SASL_CONTINUE: i32 = 11;
    pub const SASL_FINAL: i32 = 12;
}

// ─── Startup Message Types ──────────────────────────────────────────────────

/// What the client sent as its first message.
pub enum StartupType {
    /// SSLRequest — client wants to negotiate TLS.
    SslRequest,
    /// CancelRequest — client wants to cancel a query.
    CancelRequest,
    /// Normal StartupMessage with parameters.
    Startup(StartupMessage),
}

/// Parsed StartupMessage parameters.
pub struct StartupMessage {
    pub params: HashMap<String, String>,
}

// ─── Backend Message ────────────────────────────────────────────────────────

/// A complete message from the Postgres backend.
pub struct BackendMessage {
    /// Message type byte (e.g., b'R' for Authentication)
    pub msg_type: u8,
    /// Complete raw bytes including type and length (for forwarding)
    pub raw: BytesMut,
    /// Payload after the length field
    pub payload: BytesMut,
}

impl BackendMessage {
    /// Is this AuthenticationOk?
    pub fn is_auth_ok(&self) -> bool {
        self.msg_type == backend::AUTHENTICATION
            && self.payload.len() >= 4
            && (&self.payload[..4]).get_i32() == auth::OK
    }

    /// Is this an auth challenge that expects a client response?
    /// (Not AuthOk or SASLFinal, which require no client reply.)
    pub fn is_auth_challenge(&self) -> bool {
        if self.msg_type != backend::AUTHENTICATION || self.payload.len() < 4 {
            return false;
        }
        let subtype = i32::from_be_bytes([
            self.payload[0],
            self.payload[1],
            self.payload[2],
            self.payload[3],
        ]);
        subtype != auth::OK && subtype != auth::SASL_FINAL
    }

    /// Is this ReadyForQuery?
    pub fn is_ready_for_query(&self) -> bool {
        self.msg_type == backend::READY_FOR_QUERY
    }

    /// Is this ErrorResponse?
    pub fn is_error_response(&self) -> bool {
        self.msg_type == backend::ERROR_RESPONSE
    }

    /// Is this ParameterStatus?
    pub fn is_parameter_status(&self) -> bool {
        self.msg_type == backend::PARAMETER_STATUS
    }

    /// Is this BackendKeyData?
    pub fn is_backend_key_data(&self) -> bool {
        self.msg_type == backend::BACKEND_KEY_DATA
    }

    /// Is this RowDescription?
    #[allow(dead_code)]
    pub fn is_row_description(&self) -> bool {
        self.msg_type == backend::ROW_DESCRIPTION
    }

    /// Is this DataRow?
    #[allow(dead_code)]
    pub fn is_data_row(&self) -> bool {
        self.msg_type == backend::DATA_ROW
    }

    /// Return the auth subtype (e.g. OK=0, CLEARTEXT=3, MD5=5, SASL=10).
    /// Returns `None` if this isn't an Authentication message or payload is too short.
    pub fn auth_subtype(&self) -> Option<i32> {
        if self.msg_type != backend::AUTHENTICATION || self.payload.len() < 4 {
            return None;
        }
        Some(i32::from_be_bytes([
            self.payload[0],
            self.payload[1],
            self.payload[2],
            self.payload[3],
        ]))
    }

    /// Extract human-readable error message from an ErrorResponse.
    pub fn error_message(&self) -> String {
        if !self.is_error_response() {
            return String::from("not an error");
        }
        let mut parts = Vec::new();
        let mut offset = 0;
        let data = &self.payload;

        while offset < data.len() {
            let field_type = data[offset];
            if field_type == 0 {
                break;
            }
            offset += 1;

            // Find null terminator
            let str_end = data[offset..]
                .iter()
                .position(|&b| b == 0)
                .map(|p| offset + p)
                .unwrap_or(data.len());

            let value = String::from_utf8_lossy(&data[offset..str_end]).to_string();
            offset = str_end + 1;

            match field_type {
                b'M' => parts.insert(0, value), // Message
                b'D' => parts.push(value),       // Detail
                _ => {}
            }
        }

        if parts.is_empty() {
            String::from("unknown error")
        } else {
            parts.join(": ")
        }
    }
}

// ─── Parsing ────────────────────────────────────────────────────────────────

/// Try to read a complete startup-phase message from the buffer.
///
/// Startup messages have no type byte — they start with Int32 length.
/// Returns `None` if not enough data. Consumes the message from `buf` on success.
pub fn try_read_startup(buf: &mut BytesMut) -> Option<StartupType> {
    if buf.len() < 8 {
        return None;
    }

    let length = i32::from_be_bytes([buf[0], buf[1], buf[2], buf[3]]) as usize;
    if length < 8 || length > 10240 {
        return None; // sanity check
    }
    if buf.len() < length {
        return None; // need more data
    }

    let msg_buf = buf.split_to(length);
    let version = i32::from_be_bytes([msg_buf[4], msg_buf[5], msg_buf[6], msg_buf[7]]);

    match version {
        v if v == SSL_REQUEST_CODE => Some(StartupType::SslRequest),
        v if v == CANCEL_REQUEST_CODE => Some(StartupType::CancelRequest),
        _ => {
            // Parse key-value pairs
            let mut params = HashMap::new();
            let mut offset = 8;

            while offset < length - 1 {
                // Read key
                let key_end = msg_buf[offset..]
                    .iter()
                    .position(|&b| b == 0)
                    .map(|p| offset + p);
                let Some(key_end) = key_end else { break };
                let key = String::from_utf8_lossy(&msg_buf[offset..key_end]).to_string();
                offset = key_end + 1;

                // Read value
                let val_end = msg_buf[offset..]
                    .iter()
                    .position(|&b| b == 0)
                    .map(|p| offset + p);
                let Some(val_end) = val_end else { break };
                let value = String::from_utf8_lossy(&msg_buf[offset..val_end]).to_string();
                offset = val_end + 1;

                if !key.is_empty() {
                    params.insert(key, value);
                }
            }

            Some(StartupType::Startup(StartupMessage { params }))
        }
    }
}

/// Try to read a complete backend message from the buffer.
///
/// Backend messages: `u8 type | i32 length | payload`
/// Length includes itself (4 bytes) but not the type byte.
/// Returns `None` if not enough data. Consumes the message from `buf` on success.
pub fn try_read_backend_message(buf: &mut BytesMut) -> Option<BackendMessage> {
    if buf.len() < 5 {
        return None;
    }

    let msg_type = buf[0];
    let length = i32::from_be_bytes([buf[1], buf[2], buf[3], buf[4]]) as usize;
    let total_length = 1 + length; // type byte + length value

    if buf.len() < total_length {
        return None;
    }

    let raw = BytesMut::from(&buf[..total_length]);
    let payload = BytesMut::from(&buf[5..total_length]);
    buf.advance(total_length);

    Some(BackendMessage {
        msg_type,
        raw,
        payload,
    })
}

// ─── Building ───────────────────────────────────────────────────────────────

/// Build a StartupMessage with the given parameters.
pub fn build_startup_message(params: &HashMap<String, String>) -> BytesMut {
    // Calculate size: 4 (length) + 4 (version) + key-value pairs + terminal null
    let mut data_len = 4; // version
    for (key, value) in params {
        data_len += key.len() + 1 + value.len() + 1;
    }
    data_len += 1; // terminal null

    let total_len = 4 + data_len; // 4 for length field
    let mut buf = BytesMut::with_capacity(total_len);

    buf.put_i32(total_len as i32);
    buf.put_i32(PROTOCOL_VERSION_30);

    for (key, value) in params {
        buf.put_slice(key.as_bytes());
        buf.put_u8(0);
        buf.put_slice(value.as_bytes());
        buf.put_u8(0);
    }
    buf.put_u8(0); // terminal null

    buf
}

/// Build a SimpleQuery ('Q') message.
pub fn build_query_message(sql: &str) -> BytesMut {
    let msg_len = 4 + sql.len() + 1; // length field + sql + null
    let mut buf = BytesMut::with_capacity(1 + msg_len);

    buf.put_u8(b'Q');
    buf.put_i32(msg_len as i32);
    buf.put_slice(sql.as_bytes());
    buf.put_u8(0);

    buf
}

/// Build an ErrorResponse ('E') message.
pub fn build_error_response(severity: &str, sqlstate: &str, message: &str) -> BytesMut {
    let fields: Vec<(u8, &str)> = vec![
        (b'S', severity),
        (b'V', severity),
        (b'C', sqlstate),
        (b'M', message),
    ];

    // Calculate fields length
    let fields_len: usize = fields.iter().map(|(_, v)| 1 + v.len() + 1).sum::<usize>() + 1; // +1 terminal null
    let msg_len = 4 + fields_len;

    let mut buf = BytesMut::with_capacity(1 + msg_len);
    buf.put_u8(b'E');
    buf.put_i32(msg_len as i32);

    for (field_type, value) in &fields {
        buf.put_u8(*field_type);
        buf.put_slice(value.as_bytes());
        buf.put_u8(0);
    }
    buf.put_u8(0); // terminal null

    buf
}

/// Build an AuthenticationCleartextPassword request (server → client).
pub fn build_auth_cleartext_request() -> BytesMut {
    // 'R' | int32 len(8) | int32 subtype(3)
    let mut buf = BytesMut::with_capacity(9);
    buf.put_u8(backend::AUTHENTICATION);
    buf.put_i32(8); // length: 4 (len field) + 4 (subtype)
    buf.put_i32(auth::CLEARTEXT_PASSWORD);
    buf
}

/// Build an AuthenticationOk message (server → client).
pub fn build_auth_ok() -> BytesMut {
    let mut buf = BytesMut::with_capacity(9);
    buf.put_u8(backend::AUTHENTICATION);
    buf.put_i32(8);
    buf.put_i32(auth::OK);
    buf
}

/// Try to read a PasswordMessage ('p') from the client buffer.
/// Returns the password string (without null terminator) or None if not enough data.
pub fn try_read_password_message(buf: &mut BytesMut) -> Option<String> {
    if buf.len() < 5 {
        return None;
    }
    if buf[0] != b'p' {
        return None;
    }
    let length = i32::from_be_bytes([buf[1], buf[2], buf[3], buf[4]]) as usize;
    let total = 1 + length;
    if buf.len() < total {
        return None;
    }
    // Password is between offset 5 and total-1 (strip null terminator)
    let password_end = if total > 5 && buf[total - 1] == 0 {
        total - 1
    } else {
        total
    };
    let password = String::from_utf8_lossy(&buf[5..password_end]).to_string();
    buf.advance(total);
    Some(password)
}

/// Build a PasswordMessage ('p') for sending to the server.
pub fn build_password_message(password: &[u8]) -> BytesMut {
    let msg_len = 4 + password.len() + 1; // length field + password + null
    let mut buf = BytesMut::with_capacity(1 + msg_len);
    buf.put_u8(b'p');
    buf.put_i32(msg_len as i32);
    buf.put_slice(password);
    buf.put_u8(0);
    buf
}

/// Build a SASLInitialResponse message ('p') with mechanism name and initial data.
pub fn build_sasl_initial_response(mechanism: &str, data: &[u8]) -> BytesMut {
    // 'p' | int32 len | mechanism\0 | int32 data_len | data
    let msg_len = 4 + mechanism.len() + 1 + 4 + data.len();
    let mut buf = BytesMut::with_capacity(1 + msg_len);
    buf.put_u8(b'p');
    buf.put_i32(msg_len as i32);
    buf.put_slice(mechanism.as_bytes());
    buf.put_u8(0);
    buf.put_i32(data.len() as i32);
    buf.put_slice(data);
    buf
}

/// Build a SASLResponse message ('p') with response data.
pub fn build_sasl_response(data: &[u8]) -> BytesMut {
    let msg_len = 4 + data.len();
    let mut buf = BytesMut::with_capacity(1 + msg_len);
    buf.put_u8(b'p');
    buf.put_i32(msg_len as i32);
    buf.put_slice(data);
    buf
}

// ─── SQL Escaping ───────────────────────────────────────────────────────────

/// Escape a value as a SQL single-quoted literal.
/// Rejects characters that have no business in a tenant ID.
#[allow(dead_code)]
pub fn escape_literal(value: &str) -> io::Result<String> {
    if !value
        .chars()
        .all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '-' || c == '.')
    {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            format!("invalid tenant ID: disallowed characters in '{value}'"),
        ));
    }
    Ok(format!("'{}'", value.replace('\'', "''")))
}

/// Escape a value as a SQL single-quoted literal for SET commands.
///
/// Unlike `escape_literal()` which restricts characters (for untrusted tenant IDs),
/// this allows any content (resolver results come from the database and may contain
/// array literals like `{a,b,c}`, commas, spaces, etc.). Defense is quote-doubling only.
pub fn escape_set_value(value: &str) -> String {
    format!("'{}'", value.replace('\'', "''"))
}

/// Quote an identifier (double-quoted).
pub fn quote_ident(value: &str) -> io::Result<String> {
    if !value
        .chars()
        .all(|c| c.is_ascii_alphanumeric() || c == '_')
    {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            format!("invalid identifier: '{value}'"),
        ));
    }
    Ok(format!("\"{}\"", value.replace('"', "\"\"")))
}
